plugins/dfa-interpreter.ts

```diff
@@ -1,95 +1,207 @@
-declare var Interpreter: any, InterpreterGraph: any, Program: any, ProgramInput: any, ProgramOutput: any, RunningProgram: any;
-//import { Graph, Node, Edge } from "../app/models/plugin";
+/**
+ * Types which can be used for program input.
+ */
+export type ProgramInput = string;
+/**
+ * Types which can be used for program output.
+ */
+export type ProgramOutput = string | boolean;
+/**
+ * Represents a function which converts graphs into programs or else returns an error that has toString.
+ */
+export type Interpreter = (graph: Graph) => Program;
 
-class Node {
-    isAcceptState: boolean;
-    isStartState: boolean;
-    children: Edge[];
+/**
+ * This interface is to be used for debugging support and is expected to maintain mutable state.
+ */
+export interface RunningProgram {
+    /**
+     * These are the properties that will be displayed to the user. TODO: Allow for this to be more flexible if there are large numbers of properties available.
+     */
+    debugProperties: [string];
+
+    /**
+     * Gets the result of the computation after the program completes. Behavior is undefined if called when isComplete is false.
+     */
+    result: ProgramOutput | null;
+
+    /**
+     * Performs one unit of work in the forward direction. Advanced debugging support should be provided elsewhere (such as step over or continue).
+     */
+    step(): void;
+
+    /**
+     * Performs one unit of work backwards. This method is optional since backwards debugging may be non-trivial for some plugins.
+     */
+    stepBack?(): void;
+
+    /**
+     * Retrieves the value of a property enumerated in debugProperties.
+     */
+    getDebugValue(property: string): ProgramOutput;
 }
 
-class Graph {
-    edges: Edge[];
-    nodes: Node[];
+/**
+ * This represents a compiled program given an input computation graph. It should be immutable, though the RunningProgram returned by initDebugging may be mutable itself.
+ * If desired, a simple run method or initDebugging method can be provided and then fillInProgram will fill out the rest of the required fields/methods.
+ */
+export interface Program {
```

Comment:
I the future, I think these interfaces and type aliases should be in their own dedicated file. It looks to me like these are meant to be part of the plugin API that defines the contract for what a valid Sinap interpreter should be.